### PR TITLE
Pruning a Large Git Repo - new doc

### LINKED ISSUE #1626: New doc: Pruning a Large Git Repository
## Issue Description
Need new doc to provide customers with a strategy and best practices  for pruning down a large git repository. Instructions provided by Ray:

### Step 1: Make a local clone of the pantheon repository
```
git clone ssh://codeserver.dev.482ae619-7334-4013-b8d2-01c2e765cc70@codeserver.dev.482ae619-7334-4013-b8d2-01c2e765cc70.drush.in:2222/~/repository.git
```
Change directory to the repository. (default is the name of the repository minus .git)
`cd repository`

### Step 2: We need to create local copies of all remote branches
```
for BRANCH in $(git branch -r | grep -v HEAD | grep -v master); do git branch --track "${BRANCH#origin/}"; done
```

### Step 3: Get a full list of all local branches
```
BRANCHES=$(for BRANCH in $(git branch --list | grep -v master); do echo "${BRANCH}"; done; echo master)
```

### Step 4: Get a list of all large files in all branches and add it to a file called "large_files.txt" (this may take several minutes)
```
git rev-list $BRANCHES | while read rev; do git ls-tree -lr $rev | cut -c54- | grep -v '^ '; done | sort -u | perl -e '
while (<>) {
chomp;
@stuff=split("\t");
$sums{$stuff[1]} += $stuff[0];
}
print "$sums{$_} $_\n" for (keys %sums);
' | sort -rn > large_files.txt
```

### Step 5: Determine what patterns we are going to filter out. A pattern may be the path of a file by name, the path of a directory by name, or an expandable path. Inspect large_files.txt
Example of a pattern for a single file name
`PATTERN1=myfile.txt`
Example of a pattern for a directory. This will also match on all files under that directory.
`PATTERN2=my_directory`
Example of an expandable path pattern that matches all sql files with in mydirectory
`PATTERN3=mydirectory/*.sql`

### Step 6: Filter out all files and directories that match one of our patterns (This could take hours)
```
git filter-branch --force --index-filter 'git rm -rf --cached --ignore-unmatch PATTERN1 PATTERN2 PATTERN3 ...' --prune-empty --tag-name-filter cat -- --all
```

### Step 7: Push your local changes to the remote repository
```
git push origin --force --all
git push origin --force --tags
```

### Step 8: Recover local disk space (This will only reduce the size of your local repository)
```
git for-each-ref --format='delete %(refname)' refs/original | git update-ref --stdin
git reflog expire --expire=now --all
git gc --prune=now
```

### More note's: 
Pantheon is configured to run garbage collection automatically on an as needed basis. When a garbage collection is run it will prune loose objects older then 2 weeks. Also entries in the reflog will expire after 90 days. We can run Step 8 on a codeserver repository if we need to perform this garbage collection immediately.
Until the remote repository is properly garbage collected all new clones will contain the loose objects and not shrink in size. However, by performing Step 8 on the new clone it will prune the loose objects locally again. These loose objects will only affect how long it takes to make an initial clone. They will not slow down new commits or merges since they no longer need to be referenced.
Git filter-branch may need to re-write references to tags. When new tags are pushed to the pantheon repository it may cause the test and live environments to re-sync.

For reference, see: https://getpantheon.atlassian.net/browse/BUGS-319

### PR DESCRIPTION
**Pull request by nataliejeremy**
Closes #1626 

## Remaining Work
- [ ] Work with @tinefin to add any additional information/steps.

### CODE REVIEW

**Comment by hengkit** [comment]
There's a bit that requires perl to be installed and executed interactively. We might need to have a prerequisites section.

**Comment by nataliejeremy** [comment]
@hengkit @rachelwhitton Could you both help get this one? @tinefin is requesting this asap for a customer.

**Comment by nataliejeremy** [comment]
Thanks for updating this so quickly! @hengkit 
@rachelwhitton Can you review and merge? Thanks.

File: source/_docs/pruning-git-repo.md [hunk 1]
```diff
@@ -5,35 +5,31 @@ categories: [developing]
...
-git clone ssh://codeserver.dev.482ae619-7334-4013-b8d2-01c2e765cc70@codeserver.dev.482ae619-7334-4013-b8d2-01c2e765cc70.drush.in:2222/~/repository.git
-```
+1. Follow [these instructions](https://pantheon.io/docs/git/#clone-your-site-codebase) to create your local clone
 2. Change directory to the repository by running: `cd repository-name`
 
 ### Step 2: Create local copies of all remote branches.
-For a branch in `$(git branch -r | grep -v HEAD | grep -v master)`, run:
 ```
-git branch --track "${BRANCH#origin/}"; done
+for BRANCH in `git branch -r | grep -v HEAD | grep -v master`; do git branch --track "${BRANCH#origin/}"; done
```
Comment (rachelwhitton) [inline source/_docs/pruning-git-repo.md]: @hengkit: What should the local branches be tracking? I'm seeing all branches track local master, such as: <code>Branch xml-sitemap set up to track <strong>local branch master</strong> by rebasing.</code> Is that the intended result? 

Suggesting we use the following instead: 
```
for BRANCH in `git branch -r | grep -vE "HEAD|master"`; do git branch --track ${BRANCH#origin/} $BRANCH; done
```
Produces: <code>Branch xml-sitemap set up to track <strong>remote branch xml-sitemap from origin</strong> by rebasing.</code> which I think is what we’d want, yeah?

Comment (hengkit) [inline source/_docs/pruning-git-repo.md]: In your case, it looks like you already had the xml-sitemap branch locally. I didn't test for that. I concede to your logic.

**Comment by rachelwhitton** [comment]
+1 thanks @hdahme!

@nataliejeremy can you do a final review? then it's ready to 🚢 :shipit:

**Pull request merged**